deployment/params.py

```diff
@@ -242,66 +122,156 @@ def _validate_transaction_args(args: typing.Tuple[Any, ...], abi) -> typing.Dict
+            self._validate_constructor_abi_inputs(
+                contract_name=contract,
+                abi_inputs=contract_container.constructor.abi.inputs,
+                parameters=parameters,
+            )
+        
+    def _validate_constructor_param(self, param: Any, contracts: List[str]) -> None:
+        """Validates a single constructor parameter or a list of parameters."""
+        if isinstance(param, list):
+            for p in param:
+                self._validate_constructor_param(p, contracts)
+            return
+        
+        if not _is_variable(param):
+            return  # literally a value
+        variable = param.strip(VARIABLE_PREFIX)
+
+        if _is_proxy_variable(variable):
+            proxy, target = variable.split(SPECIAL_VARIABLE_DELIMITER)
+            if proxy != PROXY_NAME:
+                raise ConstructorParameters.Invalid(
+                    f"Ambiguous proxy variable {param}; only {PROXY_NAME} "
+                    f"allowed before '{SPECIAL_VARIABLE_DELIMITER}'"
+                )
+            variable = target  # check proxy target
+
+        if _is_special_variable(variable):
+            return  # special variables are always marked as valid
+
+        if variable in contracts:
+            return
+
+        raise ConstructorParameters.Invalid(f"Variable {param} is not resolvable")
+    
+    def _validate_constructor_abi_inputs(
+        self, contract_name: str, abi_inputs: List[Any], parameters: OrderedDict
+    ) -> None:
+        """Validates the constructor parameters against the constructor ABI."""
+        if len(parameters) != len(abi_inputs):
+            raise ConstructorParameters.Invalid(
+                f"Constructor parameters length mismatch - "
+                f"{contract_name} ABI requires {len(abi_inputs)}, Got {len(parameters)}."
+            )
+        if not abi_inputs:
+            return  # no constructor parameters
+
+        codex = enumerate(zip(abi_inputs, parameters.items()), start=0)
+        for position, (abi_input, resolved_input) in codex:
+            name, value = resolved_input
+            # validate name
+            if abi_input.name != name:
+                raise ConstructorParameters.Invalid(
+                    f"{contract_name} constructor parameter '{name}' at position {position} does not "
+                    f"match the expected ABI name '{abi_input.name}'."
+                )
+
+            # validate value type
+            value_to_validate = value
+            if _is_variable(value):
+                value_to_validate = self._resolve_param(value)
+            if not w3.is_encodable(abi_input.type, value_to_validate):
+                raise ConstructorParameters.Invalid(
+                    f"Constructor param name '{name}' at position {position} has a value '{value}' "
+                    f"whose type does not match expected ABI type '{abi_input.type}'"
+                )
```

Comment:
many of (all?) of these methods are stateless and do not need access to the class's state -- consider relocating them to a helper module of pure functions to keep this class small, improve modularity and unit testing ease.

Reply:
Not really. Once I created `_resolve_constants` to be stateful (as it uses constants defined in the config), then practically all the methods refactored here had dependencies going up. I tried to move as less things as possible.

Reply:
The trade off of injecting `self.constants[name]` into `_resolve_constants` as a parameter input to decouple all these methods from the class is a good one imo.

Reply:
`_resolve_constants` => `_resolve_special_variable` => `_resolve_param` => `_validate_constructor_abi_inputs`
 => `validate_constructor_parameters`

Reply:
Precisely -- if you inject into `self.constants[name]` into `_resolve_constants` instead of using instance state, the rest of the methods can be relocated as helper functions too.

Reply:
Addressed in 9b23df2b1e1be0870a4e8a2f4e93d4b68a384a1d (Thanks @KPrasch 🙏 )